### PR TITLE
Elementwise bounds fixed

### PR DESCRIPTION
**Pull request by peilinrao**
Now cartpole can use different limits.

### CODE REVIEW

**Review by Trinkle23897 — COMMENTED** [review]
The final goal is to make sure when we call `env.action_space.high` or `env.observation_space.low` on the Python side, the result is correct. Have you checked that?

```python
In [4]: env = gym.make("CartPole-v1")

In [5]: env.observation_space.high
Out[5]: 
array([4.8000002e+00, 3.4028235e+38, 4.1887903e-01, 3.4028235e+38],
      dtype=float32)
```

File: envpool/core/spec.h [hunk 1]
```diff
@@ -63,14 +63,24 @@ class Spec : public ShapeSpec {
   using dtype = D;
   std::tuple<dtype, dtype> bounds = {std::numeric_limits<dtype>::min(),
                                      std::numeric_limits<dtype>::max()};
+  std::vector<std::tuple<dtype, dtype>> elementwise_bounds;
```
[the remark this answers is not in the document]
Comment (peilinrao) [reply envpool/core/spec.h]: I think now it should work. I added a new bound_test.py in /examples to show that test_envs.observation_space.high is correct.

File: envpool/core/spec.h [hunk 2]
```diff
@@ -63,14 +63,30 @@ class Spec : public ShapeSpec {
   using dtype = D;
   std::tuple<dtype, dtype> bounds = {std::numeric_limits<dtype>::min(),
                                      std::numeric_limits<dtype>::max()};
+  std::tuple<std::vector<dtype>, std::vector<dtype>> elementwise_bounds;
+  bool enable_elementwise_bounds = false;
```
Comment (Trinkle23897) [inline envpool/core/spec.h:67]: unused variable?

File: envpool/core/spec.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Trinkle23897) [reply envpool/core/spec.h]: > I think now it should work. I added a new bound_test.py in /examples to show that test_envs.observation_space.high is correct.

test should not be under `examples/`.

File: envpool/python/protocol.py [hunk 3]
```diff
@@ -90,18 +90,29 @@ def reward_threshold(self) -> Optional[float]:
...
-  def __init__(self, dtype: Type, shape: List[int], bounds: Tuple[Any, Any]):
+  def __init__(
+    self, dtype: Type, shape: List[int], bounds: Tuple[Any, Any],
+    element_wise_bounds: Tuple[Any, Any]
+  ):
     """Constructor of ArraySpec."""
     self.dtype = dtype
     self.shape = shape
     self.minimum, self.maximum = bounds
+    self.minimum_elementwise, self.maximum_elementwise = element_wise_bounds
```
Comment (Trinkle23897) [inline envpool/python/protocol.py:101]: I personally don't think we should treat them separately.

**Review by Trinkle23897 — APPROVED** [review]

**Pull request merged**